### ISSUE TITLE
Document recommendation to turn no-descending-specificity off if using a lot of nesting

### ISSUE THREAD

**Issue by lucasvazq** [stylelint/stylelint#5894]
### What steps are needed to reproduce the bug?

I think this isn't specifically a bug, but the repo doesn't have a discussion section. The problem I have is that I can't find a way to sort my scss styles. I constantly get `no-descending-specificity` and `no-duplicate-selectors` errors.

```scss
.style-custom-scroll {
  &::-webkit-scrollbar-track {
    background-color: 'yellow';
  }

  &::-webkit-scrollbar-thumb {
    &:active {
      background-color: 'red';
    }

    &:hover {
      background-color: 'blue';
    }
  }

  &:active,
  &:hover {
    &::-webkit-scrollbar-track {
      background-color: 'green';
    }
  }
}
```

This actually output `no-descending-specificity` error:
- Expected selector ".style-custom-scroll:active" to come before selector no-descending-specificity ".style-custom-scroll::-webkit-scrollbar-thumb:active"

### What Stylelint configuration is needed to reproduce the bug?

Dependencies:
```json
"postcss": "^8.4.5",
"postcss-import": "^14.0.2",
"sass": "^1.48.0",
"stylelint": "^14.2.0",
"stylelint-config-sass-guidelines": "^9.0.1",
"stylelint-config-standard": "^24.0.0",
"tailwindcss": "^3.0.14",
```

Config (package.json):
```json
{
   "extends":[
      "stylelint-config-standard",
      "stylelint-config-sass-guidelines"
   ],
   "rules":{
      "max-nesting-depth":[
         5,
         {
            "ignore":[
               "pseudo-classes"
            ]
         }
      ],
      "no-invalid-position-at-import-rule":[
         true,
         {
            "ignoreAtRules":[
               "use"
            ]
         }
      ],
      "scss/at-rule-no-unknown":[
         true,
         {
            "ignoreAtRules":[
               "tailwind",
               "apply",
               "layer",
               "variants",
               "responsive",
               "screen"
            ]
         }
      ]
   }
}
```

### How did you run Stylelint?

yarn run stylelint ./src/styles/base.scss --fix

### Which version of Stylelint are you using?

14.2.0

### What did you expect to happen?

I want to remove the errors.

### What actually happened?

I get constantly `no-descending-specificity` and `no-duplicate-selectors` errors and I cant get out of them.

### Does the bug relate to non-standard syntax?

No

**Comment by lucasvazq** [comment]
###### I put my attempts here

## 1

```scss
.style-custom-scroll {
  &::-webkit-scrollbar-track {
    background-color: 'yellow';
  }

  &::-webkit-scrollbar-thumb {
    &:active {
      background-color: 'red';
    }

    &:hover {
      background-color: 'blue';
    }
  }

  &:active,
  &:hover {
    &::-webkit-scrollbar-track {
      background-color: 'green';
    }
  }
}
/*
 16:3  ✖  Expected selector ".style-custom-scroll:active" to come before selector ".style-custom-scroll::-webkit-scrollbar-thumb:active"  no-descending-specificity
 16:3  ✖  Expected selector ".style-custom-scroll:active" to come before selector ".style-custom-scroll::-webkit-scrollbar-thumb:hover"   no-descending-specificity
 17:3  ✖  Expected selector ".style-custom-scroll:hover" to come before selector ".style-custom-scroll::-webkit-scrollbar-thumb:active"   no-descending-specificity
 17:3  ✖  Expected selector ".style-custom-scroll:hover" to come before selector ".style-custom-scroll::-webkit-scrollbar-thumb:hover"    no-descending-specificity
*/
```

## 2
```scss
.style-custom-scroll {
  &::-webkit-scrollbar-track {
    background-color: 'yellow';

    &:active,
    &:hover {
      background-color: 'green';
    }
  }

  &::-webkit-scrollbar-thumb {
    &:active {
      background-color: 'red';
    }

    &:hover {
      background-color: 'blue';
    }
  }
}
/*
 11:3  ✖  Expected selector ".style-custom-scroll::-webkit-scrollbar-thumb" to come before selector ".style-custom-scroll::-webkit-scrollbar-track:active"  no-descending-specificity
 11:3  ✖  Expected selector ".style-custom-scroll::-webkit-scrollbar-thumb" to come before selector ".style-custom-scroll::-webkit-scrollbar-track:hover"   no-descending-specificity
*/
```

## 3
```scss
.style-custom-scroll {
  &::-webkit-scrollbar-track {
    background-color: 'yellow';
  }

  &:active,
  &:hover {
    &::-webkit-scrollbar-track {
      background-color: 'green';
    }
  }

  &::-webkit-scrollbar-thumb {
    &:active {
      background-color: 'red';
    }

    &:hover {
      background-color: 'blue';
    }
  }
}
/*
 13:3  ✖  Expected selector ".style-custom-scroll::-webkit-scrollbar-thumb" to come before selector ".style-custom-scroll:active"                           no-descending-specificity
 13:3  ✖  Expected selector ".style-custom-scroll::-webkit-scrollbar-thumb" to come before selector ".style-custom-scroll:hover"                            no-descending-specificity
 13:3  ✖  Expected selector ".style-custom-scroll::-webkit-scrollbar-thumb" to come before selector ".style-custom-scroll:active::-webkit-scrollbar-track"  no-descending-specificity
 13:3  ✖  Expected selector ".style-custom-scroll::-webkit-scrollbar-thumb" to come before selector ".style-custom-scroll:hover::-webkit-scrollbar-track"   no-descending-specificity
*/
```

## 4
```scss
.style-custom-scroll {
  &::-webkit-scrollbar-track {
    background-color: 'yellow';
  }

  &:active {
    &::-webkit-scrollbar-track {
      background-color: 'green';
    }

    &::-webkit-scrollbar-thumb {
      background-color: 'red';
    }
  }

  &:hover {
    &::-webkit-scrollbar-track {
      background-color: 'green';
    }

    &::-webkit-scrollbar-thumb {
      background-color: 'blue';
    }
  }
}
/*
 14:3  ✖  Expected selector ".style-custom-scroll:hover" to come before selector ".style-custom-scroll:active::-webkit-scrollbar-track"  no-descending-specificity
 14:3  ✖  Expected selector ".style-custom-scroll:hover" to come before selector ".style-custom-scroll:active::-webkit-scrollbar-thumb"  no-descending-specificity
*/
```

## 5
```scss
.style-custom-scroll {
  &::-webkit-scrollbar-track {
    background-color: 'yellow';
  }

  &:hover {
    &::-webkit-scrollbar-track {
      background-color: 'green';
    }

    &::-webkit-scrollbar-thumb {
      background-color: 'blue';
    }
  }

  &:active {
    &::-webkit-scrollbar-track {
      background-color: 'green';
    }

    &::-webkit-scrollbar-thumb {
      background-color: 'red';
    }
  }
}
/*
 14:3  ✖  Expected selector ".style-custom-scroll:active" to come before selector ".style-custom-scroll:hover::-webkit-scrollbar-track"  no-descending-specificity
 14:3  ✖  Expected selector ".style-custom-scroll:active" to come before selector ".style-custom-scroll:hover::-webkit-scrollbar-thumb"  no-descending-specificity
*/
```

---

I appreciate any kind of help, thanks :smile: :pray:

**Comment by lucasvazq** [comment]
Just for reference, this is my original `.style-custom-scroll` class definition:
```scss
.style-custom-scroll {
  &::-webkit-scrollbar {
    width: 1rem;
    height: 0.5rem;
  }
  &::-webkit-scrollbar-thumb {
    background-color: map.get($colors, primary-focus);
    border-radius: 0.25rem;
    &:hover {
      box-shadow: 0 0 0.125rem 1px rgba(0, 0, 0, 0.2);
    }
  }
  &::-webkit-scrollbar-track {
    border-radius: 0.25rem;
  }

  &.style-scroll-top {
    &::-webkit-scrollbar-thumb {
      border-top-left-radius: 0;
      border-top-right-radius: 0;
    }
  }
  &.style-scroll-bottom {
    &::-webkit-scrollbar-thumb {
      border-bottom-left-radius: 0;
      border-bottom-right-radius: 0;
    }
  }
}
body.style-custom-scroll {
  &::-webkit-scrollbar {
    background-color: map.get($colors, secondary);
  }
  &::-webkit-scrollbar-thumb {
    &:hover {
      background-color: map.get($colors, primary);
    }
    &:active {
      background-color: color.scale(map.get($colors, primary-focus), $lightness: 15%);
    }
  }
  &::-webkit-scrollbar-track {
    background-color: map.get($colors, secondary);
    &:active, &:hover {
      background-color: color.scale(map.get($colors, secondary), $lightness: 10%);
    }
  }
}
:not(body).style-custom-scroll {
  &::-webkit-scrollbar-thumb {
    &:hover {
      background-color: color.scale(map.get($colors, primary-focus), $lightness: -5%);
    }
    &:active {
      background-color: color.scale(map.get($colors, primary-focus), $lightness: -10%);
    }
  }
  &::-webkit-scrollbar-track {
    background-color: color.scale(map.get($colors, primary), $lightness: 40%);
    &:active, &:hover {
      background-color: color.scale(map.get($colors, primary), $lightness: 50%);
    }
  }
}
```

I only leave a minified version in the issue that serves to represent a problem that I encountered when trying to format this original code.

**Comment by jeddy3** [comment]
Perhaps the following will satisfy the rule:

```scss
.style-custom-scroll {
  &::-webkit-scrollbar {
    width: 1rem;
    height: 0.5rem;
  }
  &::-webkit-scrollbar-thumb {
    background-color: map.get($colors, primary-focus);
    border-radius: 0.25rem;
    &:hover {
      box-shadow: 0 0 0.125rem 1px rgba(0, 0, 0, 0.2);
      background-color: color.scale(map.get($colors, primary-focus), $lightness: -5%);
    }
    
     &:active {
      background-color: color.scale(map.get($colors, primary-focus), $lightness: -10%);
    }
  }
  &::-webkit-scrollbar-track {
    border-radius: 0.25rem;
    background-color: color.scale(map.get($colors, primary), $lightness: 40%);
    &:active, &:hover {
      background-color: color.scale(map.get($colors, primary), $lightness: 50%);
    }
  }

  &.style-scroll-top {
    &::-webkit-scrollbar-thumb {
      border-top-left-radius: 0;
      border-top-right-radius: 0;
    }
  }
  &.style-scroll-bottom {
    &::-webkit-scrollbar-thumb {
      border-bottom-left-radius: 0;
      border-bottom-right-radius: 0;
    }
  }
}

body.style-custom-scroll {
  &::-webkit-scrollbar {
    background-color: map.get($colors, secondary);
  }
  &::-webkit-scrollbar-thumb {
    &:hover {
      background-color: map.get($colors, primary);
    }
    &:active {
      background-color: color.scale(map.get($colors, primary-focus), $lightness: 15%);
    }
  }
  &::-webkit-scrollbar-track {
    background-color: map.get($colors, secondary);
    &:active, &:hover {
      background-color: color.scale(map.get($colors, secondary), $lightness: 10%);
    }
  }
}
```

The rule [has limitations](https://stylelint.io/user-guide/rules/list/no-descending-specificity#dom-limitations) and I've opened a pull request to recommend turning it off if you use lots of nesting.

**Comment by lucasvazq** [comment]
Hi @jeddy3, thanks for the answer :1st_place_medal:

The `body.style-custom-scroll` gives me `selector-no-qualifying-type`.
Right now the problem seems very obvious to me.

I have a class called `style-custom-scroll` and I expect them have different behaviour if they are applied in a body tag or not. I think the problem here is that doing that is a bad practice. I just need to declare two or three diferent classes.

For example:
- style-scroll (this will have all the basis)
- style-scroll-primary (I will use this for body)
- style-scroll-secondary (I will use this for the rest of the app)

This is my final result:
```scss
.style-scroll {
  &::-webkit-scrollbar {
    height: 0.5rem;
    width: 1rem;
  }

  &::-webkit-scrollbar-thumb {
    background-color: map.get($colors, primary-focus);
    border-radius: 0.25rem;

    &:hover {
      box-shadow: 0 0 0.125rem 1px rgb(0 0 0 / 20%);
    }
  }

  &::-webkit-scrollbar-track {
    border-radius: 0.25rem;
  }

  &.style-scroll-top {
    &::-webkit-scrollbar-thumb {
      border-top-left-radius: 0;
      border-top-right-radius: 0;
    }
  }

  &.style-scroll-bottom {
    &::-webkit-scrollbar-thumb {
      border-bottom-left-radius: 0;
      border-bottom-right-radius: 0;
    }
  }
}

.style-scroll-primary {
  &::-webkit-scrollbar {
    background-color: map.get($colors, secondary);
  }

  &::-webkit-scrollbar-thumb {
    &:hover {
      background-color: map.get($colors, primary);
    }

    &:active {
      background-color: color.scale(map.get($colors, primary-focus), $lightness: 15%);
    }
  }

  &::-webkit-scrollbar-track {
    background-color: map.get($colors, secondary);

    &:active,
    &:hover {
      background-color: color.scale(map.get($colors, secondary), $lightness: 10%);
    }
  }
}

.style-scroll-secondary {
  &::-webkit-scrollbar-thumb {
    &:hover {
      background-color: color.scale(map.get($colors, primary-focus), $lightness: -5%);
    }

    &:active {
      background-color: color.scale(map.get($colors, primary-focus), $lightness: -10%);
    }
  }

  &::-webkit-scrollbar-track {
    background-color: color.scale(map.get($colors, primary), $lightness: 40%);

    &:active,
    &:hover {
      background-color: color.scale(map.get($colors, primary), $lightness: 50%);
    }
  }
}
```

**Comment by jeddy3** [comment]
That's a nice solution.
